PolyEngine/Engine/Src/World.hpp

```diff
@@ -86,7 +75,93 @@ namespace Poly {
+		 * @tparam SecondaryComponents Additional component types (warning: returned pointers might be null!)
+		 * @return A proxy object that can be used in a range-for loop.
+		 */
+		template<typename PrimaryComponent, typename... SecondaryComponents>
+		IteratorProxy<PrimaryComponent, SecondaryComponents...> IterateComponents()
+		{
+			return {this};
+		}
+
+		//------------------------------------------------------------------------------
+		/**
+		 * An implementation detail
+		 */
+		template<typename PrimaryComponent, typename... SecondaryComponents>
+		class ComponentIterator : public BaseObject<>,
+		                          public std::iterator<std::bidirectional_iterator_tag, std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type...>>
+		{
+			public:
+			bool operator==(const ComponentIterator& rhs) const { return primary_iter == rhs.primary_iter; }
+			bool operator!=(const ComponentIterator& rhs) const { return !(*this == rhs); }
+
+			std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type...> operator*() const
+			{
+				PrimaryComponent* primary = &*primary_iter;
+				return std::make_tuple(primary, primary->template GetSibling<SecondaryComponents>()...);
+			}
+			std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type...> operator->() const
+			{
+				return **this;
+			}
+
+			ComponentIterator& operator++() { ++primary_iter; return *this; }
+			ComponentIterator operator++(int) { ComponentIterator ret(primary_iter); ++primary_iter; return ret; }
+			ComponentIterator& operator--() { --primary_iter; return *this; }
+			ComponentIterator operator--(int) { ComponentIterator ret(primary_iter); --primary_iter; return ret; }
+
+			private:
+			explicit ComponentIterator(typename IterablePoolAllocator<PrimaryComponent>::Iterator parent) : primary_iter(parent) {}
+			friend struct IteratorProxy<PrimaryComponent, SecondaryComponents...>;
+
+			typename IterablePoolAllocator<PrimaryComponent>::Iterator primary_iter;
+		};
+
+		/**
+		 * An implementation detail
+		 */
+		template<typename PrimaryComponent, typename... SecondaryComponents>
+		struct IteratorProxy {
+			World::ComponentIterator<PrimaryComponent, SecondaryComponents...> Begin() {
+				auto alloc = static_cast<IterablePoolAllocator<PrimaryComponent>*>(w->ComponentAllocators[w->EnginePtr->GetComponentID<PrimaryComponent>()]);
+				return ComponentIterator<PrimaryComponent, SecondaryComponents...>(alloc->Begin());
+			}
+			World::ComponentIterator<PrimaryComponent, SecondaryComponents...> End() {
+				auto alloc = static_cast<IterablePoolAllocator<PrimaryComponent>*>(w->ComponentAllocators[w->EnginePtr->GetComponentID<PrimaryComponent>()]);
+				return ComponentIterator<PrimaryComponent, SecondaryComponents...>(alloc->End());
+			}
+			auto begin() { return Begin(); }
```

Comment:
Those can be removed since begin and end are provided at the bottom of the file.